src/Orleans/Threading/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Concurrency;
+
+namespace Orleans.Threading
+{
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private static readonly CancellationTokenSource CancelledTokenSource = new CancellationTokenSource(1);
+
+        private readonly ConcurrentDictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
```

Comment:
The method that is accessing this dictionary is marked with `AlwaysInterleave` attribute.